### PR TITLE
Add diagnose command

### LINKED ISSUE #412: Diagnose command
It is currently really difficult to tell where a problem is when trying to diagnose a KET cluster. 
It would be good if we had:
- [x] Service status 
- [x] Logs for all services
- [x] `docker ps -a` output from all machines
- [x] `docker logs` for the control plane and addons we install
- [x] calico diagnose info

### PR DESCRIPTION
**Pull request by dkoshkin**
Fixes #412 

This change will also run the `diagnose` command after an integration test failure.

Sample Output:
```
➜  out git:(master) ✗ ./kismatic diagnose -f kismatic-cluster-1.yaml --verbose
Reading plan file                                                               [OK]
Validate SSH connectivity to nodes                                              [OK]
Diagnose Nodes
- Running task:
  ip-10-0-0-96                                                                  [OK]
  ip-10-0-0-94                                                                  [OK]
  ip-10-0-0-17                                                                  [OK]
  ip-10-0-0-148                                                                 [OK]
- Running task: create /tmp/diagnostics-20170321_164611 directory
  ip-10-0-0-96                                                                  [OK]
  ip-10-0-0-17                                                                  [OK]
  ip-10-0-0-94                                                                  [OK]
  ip-10-0-0-148                                                                 [OK]
- Running task: diagnose etcd nodes
  ip-10-0-0-94                                                                  [SKIPPED]
  ip-10-0-0-17                                                                  [SKIPPED]
  ip-10-0-0-148                                                                 [SKIPPED]
  ip-10-0-0-96                                                                  [OK]
- Running task: diagnose master nodes
  ip-10-0-0-94                                                                  [SKIPPED]
  ip-10-0-0-17                                                                  [SKIPPED]
  ip-10-0-0-96                                                                  [SKIPPED]
  ip-10-0-0-148                                                                 [OK]
- Running task: diagnose worker nodes
  ip-10-0-0-148                                                                 [SKIPPED]
  ip-10-0-0-96                                                                  [SKIPPED]
  ip-10-0-0-17                                                                  [OK]
  ip-10-0-0-94                                                                  [OK]
- Running task: archive diagnostics directory
- Running task: copy diagnostics to local directory in /Users/dkoshkin/go/src/github.com/apprenda/kismatic/out/diagnostics/20170321_164611
  ip-10-0-0-96                                                                  [OK]
  ip-10-0-0-94                                                                  [OK]
  ip-10-0-0-17                                                                  [OK]
  ip-10-0-0-148                                                                 [OK]
```

### CODE REVIEW

File: ansible/group_vars/diagnostics.yaml [hunk 1]
```diff
@@ -0,0 +1,50 @@
+---
+diagnostics:
+  host_diagnostics:
+    - {msg: "Getting date", command: "date", file: "date"}
+    - {msg: "Getting hostname", command: "hostname", file: "hostname"}
+    - {msg: "Dumping /etc/hosts", command: "cat /etc/hosts", file: "hosts_file"}
+  docker_diagnostics:
+    - {msg: "Dumping docker.service status", command: "systemctl status docker", file: "systemd_docker"}
+    - {msg: "Dumping journal for docker.service", command: "journalctl -u docker.service --no-pager", file: "journalctl_docker"}
```
Comment (alexbrand) [inline ansible/group_vars/diagnostics.yaml]: Should we limit this to the last N entires? I actually don't know if journald rotates logs or anything, but I am worried that we'll try to ship gigabytes of logs for a long running service.

File: ansible/group_vars/diagnostics.yaml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    - {msg: "Dumping kube-controller-manager docker logs", command: "docker logs `docker ps -a -f name=k8s_kube-controller-manager --format=\\{\\{.ID\\}\\} -l`", file: "logs_kube_controller_manager"}
+    - {msg: "Dumping kube-scheduler docker logs", command: "docker logs `docker ps -a -f name=k8s_kube-scheduler --format=\\{\\{.ID\\}\\} -l`", file: "logs_kube_scheduler"}
+    - {msg: "Dumping nodes", command: "kubectl get nodes", file: "kubectl_nodes"}
+    - {msg: "Dumping apis", command: "kubectl get api-versions", file: "kubectl_apis"}
+    - {msg: "Dumping pods", command: "kubectl get pods -n kube-system", file: "kubectl_pods"}
+    - {msg: "Dumping services", command: "kubectl get services -n kube-system", file: "kubectl_services"}
+    - {msg: "Dumping daemonsets", command: "kubectl get ds -n kube-system", file: "kubectl_daemonset"}
+    - {msg: "Dumping deployments", command: "kubectl get deployments -n kube-system", file: "kubectl_deployments"}
+  k8s_worker_diagnostics:
+    - {msg: "Dumping kube-dashboard docker logs", command: "docker logs `docker ps -a -f name=k8s_kubernetes-dashboard --format=\\{\\{.ID\\}\\} -l`", file: "logs_kubernetes_dashboard"}
```
Comment (alexbrand) [inline ansible/group_vars/diagnostics.yaml]: Will we find these workloads on every worker node? I wonder if this should be `cluster_services_diagnostics`

File: ansible/group_vars/diagnostics.yaml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    - {msg: "Dumping routes (IPv6)", command: "ip -6 route", file: "ipv6_route"}
+    - {msg: "Dumping interface info (IPv4)", command: "ip -4 addr", file: "ipv4_addr"}
+    - {msg: "Dumping interface info (IPv6)", command: "ip -6 addr", file: "ipv6_addr"}
+    - {msg: "Dumping iptables (IPv4)", command: "iptables-save", file: "ipv4_tables"}
+    - {msg: "Dumping iptables (IPv6)", command: "ip6tables-save", file: "ipv6_tables"}
+    - {msg: "Dumping ipsets", command: "ipset list", file: "ipsets"}
+  etcd_diagnostics:
+    - {msg: "Dumping etcd_k8s.service status", command: "systemctl status etcd_k8s", file: "systemd_etcd_k8s"}
+    - {msg: "Dumping journal for etcd_k8s.service", command: "journalctl -u etcd_k8s.service --no-pager", file: "journalctl_etcd_k8s"}
+    - {msg: "Getting etcd_networking health", command: "/usr/bin/etcdctl --endpoint='https://127.0.0.1:6666/' --cert-file=/etc/etcd_networking/etcd.pem --key-file=/etc/etcd_networking/etcd-key.pem --ca-file=/etc/etcd_networking/ca.pem cluster-health", file: "etcd_k8s_health"}
```
Comment (alexbrand) [inline ansible/group_vars/diagnostics.yaml]: Should this be etcd_k8s?

File: integration/upgrade_test.go [hunk 4]
```diff
@@ -243,11 +243,14 @@ func extractCurrentKismaticInstaller() {
 }
 func upgradeCluster() {
 	// Perform upgrade
-	cmd := exec.Command("./kismatic", "upgrade", "offline", "-f", "kismatic-testing.yaml")
+	cmd := exec.Command("./kismatic", "upgrade", "offline", "-f", "kismatic-testing.yaml", "--skip-preflight")
```
Comment (alexbrand) [inline integration/upgrade_test.go]: Is the `--skip-preflight` intentional?

File: pkg/cli/diagnose.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+			if len(args) != 0 {
+				return fmt.Errorf("Unexpected args: %v", args)
+			}
+
+			return doDiagnostics(out, opts)
+		},
+	}
+
+	// PersistentFlags
+	cmd.PersistentFlags().StringVarP(&opts.planFilename, "plan-file", "f", "kismatic-cluster.yaml", "path to the installation plan file")
```
Comment (alexbrand) [inline pkg/cli/diagnose.go]: See https://github.com/apprenda/kismatic/blob/master/pkg/cli/install.go#L34

File: pkg/cli/diagnose.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+	// Validate SSH connectivity to nodes
+	if ok, errs := install.ValidatePlanSSHConnections(plan); !ok {
+		util.PrettyPrintErr(out, "Validate SSH connectivity to nodes")
+		util.PrintValidationErrors(out, errs)
+		return fmt.Errorf("SSH connectivity validation errors found")
+	}
+	util.PrettyPrintOk(out, "Validate SSH connectivity to nodes")
+
+	// Get versions as only supported nodes are >1.3
```
Comment (alexbrand) [inline pkg/cli/diagnose.go]: This is interesting. Does this mean we'll need version specific diagnose plays/tasks?

File: pkg/install/execute.go [hunk 7]
```diff
@@ -543,6 +583,27 @@ func (ae *ansibleExecutor) UpgradeClusterServices(plan Plan) error {
...
 }
 
+func (ae *ansibleExecutor) DiagnoseNodes(plan Plan) error {
+	inventory := buildInventoryFromPlan(&plan)
+	cc, err := ae.buildClusterCatalog(&plan)
+	if err != nil {
+		return err
+	}
+	// dateTime will be appended to the diagnostics directory
+	now := time.Now().Format("20060102_150405")
```
Comment (alexbrand) [inline pkg/install/execute.go]: Can we use the same date formatting we use for the runs directory?

File: ansible/group_vars/diagnostics.yaml (hunk 2 above)
Comment (dkoshkin) [inline ansible/group_vars/diagnostics.yaml]: It won't be unfortunately we won't know where the service is running. I wanted to keep it generic and not use `kubectl` to handle scenarios if the API server is down or some other issues with it.

File: ansible/group_vars/diagnostics.yaml (hunk 1 above)
Comment (dkoshkin) [inline ansible/group_vars/diagnostics.yaml]: Not sure, I think we should wait on this until we get some issues about it, its too hard to tell now what would be the "right" amount.

**Comment by alexbrand** [comment]
LGTM

**Pull request merged**